### PR TITLE
Make KeyValueSchema an interface visible in the public Schema API

### PR DESCRIPTION
**Pull request by eolivelli**
- allow users of pulsar-client-api to use KeyValueSchema
- move KeyValueSchema implementation to KeyValueSchemaImpl
- introduce a new interface org.apache.pulsar.client.api.schema.KeyValueSchema

### Verifying this change

The integration test file TestGenericObjectSink has been changed to use the new API class.
See the comments in https://github.com/apache/pulsar/pull/10878

### CODE REVIEW

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/KeyValueSchema.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     * @return the Schema of the Key
+     */
+    Schema<K> getKeySchema();
+
+    /**
+     * Get the Schema of the Value.
+     *
+     * @return the Schema of the Value
+     */
+    Schema<K> getValueSchema();
```
Comment (lhotari) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/KeyValueSchema.java]: Should this be `Schema<V>`?

Comment (eolivelli) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/KeyValueSchema.java]: yes, fixed

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java [hunk 2]
```diff
@@ -364,8 +364,8 @@ public void write(Record<T> record) {
         SinkRecord<T> sinkRecord = (SinkRecord<T>) record;
         TypedMessageBuilder<T> msg = pulsarSinkProcessor.newMessage(sinkRecord);
 
-        if (record.getKey().isPresent() && !(record.getSchema() instanceof KeyValueSchema &&
-                ((KeyValueSchema) record.getSchema()).getKeyValueEncodingType() == KeyValueEncodingType.SEPARATED)) {
+        if (record.getKey().isPresent() && !(record.getSchema() instanceof KeyValueSchemaImpl &&
```
Comment (lhotari) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java:367]: would it make sense to use the interface `KeyValueSchema` in `instanceof` checks?

Comment (eolivelli) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java:367]: makes sense

**Review by lhotari — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Review by sijie — APPROVED** [review]

**Pull request merged**